### ISSUE TITLE
RISC-V: Use CLINT.mtime instead of rdtime

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#942]
At least for the platforms that emulate `rdtime` in SBI. See also [this comment](https://github.com/seL4/seL4/pull/863#issuecomment-1150427868) by @axel-h.

E.g. on HIFIVE `rdtime` can take more than one microsecond, which is a hit taken on every context switch.

I don't have RISC-V hardware, just putting this issue up here so someone who cares about RISC-V performance can pick this up.

**Comment by Ivan-Velickovic** [comment]
So in the case of the HiFive Unleashed which has a U54-MC, using the `mtime` register would involve mapping the CLINT into the kernel's address space (having it as a kernel device?) and then just changing `riscv_read_time` to use `mtime` without going to SBI?

This makes me wonder why isn't the CLINT already a kernel device as is the PLIC? I doubt user-space is supposed to access anything in the CLINT?

**Comment by Ivan-Velickovic** [comment]
The Polarfire also has a U54-MC, the QEMU RISC-V virt platform seems to implement the same CLINT. Not sure about the other supported RISC-V platforms.

**Comment by Ivan-Velickovic** [comment]
By "seems to implement the same CLINT" I mean that they all have the `"riscv,cpu-intc"` compatible string, and I've just checked all the other RISC-V platforms and they also have the same compatible string. Whether having the same compatible string results in the same memory map is not something I know.

**Comment by Indanz** [comment]
Seems like the right approach. Only thing I'm not sure about is whether seL4 needs to configure anything. Presumably that's already done by the SBI or bootloader.

According to [this](https://www.rtos.org/FreeRTOS_Support_Forum_Archive/August_2019/freertos_CLINT_vs_mtime_in_FreeRTOS_RISC-V_port_2e8d6a0c4fj.html) FreeRTOS post, `mtime` and `mtimecmp` registers are standard RISC-V registers, it's just their location that differs per platform. So you can count on all RISC-V ports having it.

However, in the future I would expect native hardware support for the `rdtime` instruction, and if it's available it's always better to use that. So retain the `rdtime` implementation, but make it a platform config option.

User space has to use `rdtime`, you can't give it access without giving it the power to mess up timer interrupts for the kernel. As it seems normal device periphery memory, you could give user space access to it.

The CLINT specifications [here](https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc) might be useful too.

**Comment by Indanz** [comment]
> This seems concerning since then couldn't any user process that has the power to map in device memory mess up the kernel state?

Actually, I suppose you could give user space a read-only mapping and then it should be safe I think.

**Comment by lsf37** [comment]
> > you can't give it access without giving it the power to mess up timer interrupts for the kernel. As it seems normal device periphery memory, you could give user space access to it.
> 
> This seems concerning since then couldn't any user process that has the power to map in device memory mess up the kernel state?

Usually we do not allow users any mappings to devices that are under kernel control. Even giving read-only mappings would likely completely destroy information flow guarantees.

**Comment by Indanz** [comment]
@lsf37:
> just being able to read any shared resource uncontrolled would have to count as an information flow between any two domains, at least if the behaviour of one domain can (indirectly) influence the values there, which with timing it can.

Of course, in this case it's the `mtimecmp` register that would directly leak kernel internal information to user space. On single core it would only tell the task how much budget it has left, on multicore it tells how much budget tasks on other cores have left. On non-MCS it's useless. But it's very limited information flow that's probably not very interesting or useful.

Compared to using the cache as side channel this timing info is kind of insignificant. In other words, if you're at that level of paranoia, you shouldn't use systems with caches shared between separate domains to begin with.

**Comment by lsf37** [comment]
> Compared to using the cache as side channel this timing info is kind of insignificant. 

The amount doesn't really matter for the information flow proofs. If the user can directly read information from that page, it would be an overt channel, not a side channel (even if you can only "write" the value via timing behaviour).

Not all systems need those proofs, but that doesn't mean we can make them impossible.

> In other words, if you're at that level of paranoia, you shouldn't use systems with caches shared between separate domains to begin with.

That is correct, but exactly that feature (elimination of timing channels via cache) is in development.

**Comment by lsf37** [comment]
> This is getting off-topic, but last remark.

Almost last ;-). You're not wrong in terms of actual information flow. My point was much more technical -- we have proofs of information flow enforcement for the kernel (for AArch32 and RISC-V). These proofs are just about overt information flow, i.e. that the kernel does not accidentally hand out information it shouldn't and can't be tricked by any sequences of actions to hand out such information (e.g. read a bunch of bytes from a page and just write it to another user or forget to clean it or things like that). This is to make sure that it's even worth it to mitigate side channels and covert channels.

The proof of this weaker property is possibilistic, that is, any overt flow whatsoever, even if it is just one bit would be a violation (so in the overt-flow sense the property is very strong, but it only applies to overt flows visible in the model). This is inherent in the property that is being proved, which means we can't relax it for special situations.

Currently, the problem we have here is not visible in the model, because it is underneath the machine interface, but it violates one of the assumptions of the proofs. In fact, the current version is completely catastrophic and violates even the assumptions of functional correctness, i.e. the kernel can crash because this can be mapped. But let's assume we fix that and map it read-only, then it would only violate the explicit assumptions of the infoflow proof. If that violation becomes visible in any other part of the model (e.g. in user behaviour where that page is mapped) the entire proof would become invalid because it introduces a contradiction and everything derived from it is worthless. We can't risk that.

So while I agree with you that there are larger channels to mitigate, the proofs will insist on this one being closed completely just so we can prove a (in some sense) weaker information flow property about other parts of the kernel.

**Comment by lsf37** [comment]
Sorry to keep harping on about this, I just want to make sure I get this point across, because it's important for more than this specific instance:

> If in a configuration bad user space has the power to destroy everything, it doesn't mean that seL4's model or proofs are suddenly useless or pointless. 

That's not quite my point. My point is: if I can derive a contradiction in the model (i.e. if I can prove `False`), then I can prove anything, and that *does* make the proof worthless. If the assumptions are such that this can't happen, then it's fine, it's just an assumption about how reality looks like. The main thing is that we don't introduce inconsistency. 

> and even then bad DTS files can break those assumptions (like having multiple memory regions pointing to the same RAM via different physical addresses, which is possible sometimes).

Yes, a bad dts file can violate all kinds of proof assumptions, e.g. the one being fixed in https://github.com/seL4/seL4/pull/1057. That one is what I meant with catastrophic, because the proofs pass fine yet the kernel crashes if you map the page and write to it, because it basically makes the proofs work for a different reality.

While this can lead to a crash, this is still the "benign" case (like DMA): as long as it doesn't happen (it's not mapped or it's not written to), the proof predictions do hold, there is no internal contradiction. But if we made it part of the model that this is a timer device, and at the same time allow it to be mapped and treated formally as normal memory in user space (which is what would happen in the current model if we just allowed it to be mapped read-only), then it would become a contradiction in the model and the proof predictions would not hold any more, no matter if the feature is actually used or not, because the contradiction is a possible execution in the model.

If we really wanted to make this read-only mapping happen, we'd need to rework the main implicit assumption that there are no shared devices between user space and kernel space and introduce the concept that you can map things the kernel sees and modifies and that some or all of that is visible to user space for reading and/or writing. That could theoretically be done, it's not fundamentally impossible, but it'd be a bit of a project. Once that is done, we could indeed explicitly exclude systems that map the page from the infoflow proofs for instance. It's mainly the current "naive" model that would lead to inconsistency if we just mapped the page.

**Comment by kent-mcleod** [comment]
I ask because in another scenario (on Aarch64) there's other platforms where some MMIO register accesses cause an external abort when accessed from EL0 but not from EL1. One approach of resolving could be to add a capability that allowed a TCB to operate in EL1 instead of EL0 (at the same privilege level as the kernel). This is essentially the same authority that would be conveyed by the proposed SMC capabilities (as they allow non-kernel code to run at an even higher privilege level than the kernel). But would this also more severely break the proofs because it'd also create contradictions in the model? How is the SMC caps different?

**Comment by lsf37** [comment]
> Are you able to describe roughly what the specific contradiction would be in this case? Is it that kernel device memory isn't accessible from user level? But if it was accessible from user level you could use both facts to prove false by contradiction?

This is all if we did this naively trying to shoehorn it into the current model: my main concern would be that it behaves as a modelled device on the kernel side (timer device as it is now) and as plain memory on the user side, which is not equal, but we are implicitly treating it as equal.

Memory is the only thing users can currently map (that's where the current model is naive), device interaction in the kernel is exclusively via machine interface functions, and we prove that kernel-level mapped device memory is never accessed outside those functions (because we need to assume that all other kernel interactions with memory are actual memory). So from that you'd currently either get that a machine op changes memory the kernel can otherwise legally access, which currently would be a contradiction or you would get that the kernel changes device state you want the user to see, but it doesn't show up on the user side. The latter option might not actually be a formal contradiction -- we'd have to look into wether we can derive one from it or not. You'd then have a model that always claims that the content of the user page is unchanged, even though in reality it does change. This could be a "benign" wrong assumption for access control and infoflow where the access control theorem says the value stays the same for the user, but in reality it does not (same as for actual user-level devices at the moment).

So there *might* be a path out of it that does not introduce inconsistency in the current model, but you'd explicitly not be modelling the effect you want the user to see. The "proper" way out would be to introduce real memory-mapped devices for users (and maybe the kernel, not sure if that is necessary, would have to look into it). Those devices then would be executing concurrently with users (and kernel), because the kernel doesn't have the power to schedule them. In a formal multicore setting this could be modelled relatively nicely. In the sequential setting it'd be a hack, but potentially still possible to model at least the effects of devices visible to the user.

> I ask because in another scenario (on Aarch64) there's other platforms where some MMIO register accesses cause an external abort when accessed from EL0 but not from EL1. One approach of resolving could be to add a capability that allowed a TCB to operate in EL1 instead of EL0 (at the same privilege level as the kernel). This is essentially the same authority that would be conveyed by the proposed SMC capabilities (as they allow non-kernel code to run at an even higher privilege level than the kernel). But would this also more severely break the proofs because it'd also create contradictions in the model? How is the SMC caps different?

If those threads would fully operate in EL1 like the kernel that'd break the proofs completely, because the proofs model what part of the state users can affect and *the* major assumption there is that users don't have privileged access to memory and device state the kernel cares about. It's what makes everything work, including basic functional correctness. If the threads would operate in EL1-equivalent only for this register access privilege then maybe we could try to describe what effects this can have, assume the worst from user behaviour, and see what properties still hold.

For SMC: you'll notice in the SMC discussion that I said we don't know what to do about those in verification and they will likely remain unverified for now. I think the approach there would be to have a model of what each call is supposed to do and then assume that it does that. This is why I wanted to be able to statically (at compile time) restrict the set of possible SMC calls, so that we can have a model for "harmless" ones and don't have to introduce an assumption on system configuration for just functional correctness.

It then depends on what the SMC calls do. If it is things like setting CPU frequency or power state or other parameters that are like kernel devices, then I think it'd fit into the current model. We'd treat them like hardware from the kernel perspective (in particular assume that they cannot change model/behaviour at runtime, which is different to having EL1 user threads). If they do different things than the model, the proof predictions don't hold when they do that, but do hold when you don't use them.

If they write to user-visible memory in a way that is similar to the discussion here and we do want that effect in the model then we'd have to rework the model to introduce that as a consistent concept. That's not impossible, but depending on the details it could be a lot of work.

What you can't have is genuinely unknown privileged behaviour in the model. (I mean, yes, you can, but then you can't prove anything at all)

**Comment by kent-mcleod** [comment]
> Of course, in this case it's the `mtimecmp` register that would directly leak kernel internal information to user space. On single core it would only tell the task how much budget it has left, on multicore it tells how much budget tasks on other cores have left. On non-MCS it's useless. But it's very limited information flow that's probably not very interesting or useful.

It looks like mtimecmp is on a different 4k page from mtime on sifive anyway.
